### PR TITLE
[Needs someone to complete] Reduce sum on many axes

### LINKED ISSUE #2006: Tensor.sum() over multiple axes
During writing `einsum` I've found that `sum` is able to sum only over single axis. I'm going to add a simple implementation of it:

```
def sum(input, axes, keepdim=False):
    # probably some check for uniqueness of axes
    if keepdim:
        for ax in axes:
            input = input.sum(ax, keepdim=True)
    else:
        for ax in sorted(axes, reverse=True):
            input = input.sum(ax)
    return input
```
1. Is it ok?
2. Where should it be better placed?

### PR DESCRIPTION
**Pull request by vlasenkov**
Resolves #2006 

- [x] arrange algorithm
- [ ] arrange places to insert the feature (as separate function or as Tensor/Variable method)
- [ ] implement the feature
- [ ] rename all `keepdim` to `keepdims`
- [ ] write docs
- [ ] write tests

### CODE REVIEW

File: torch/functional.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from ._utils import _range
 from operator import mul
 from functools import reduce
+import collections
```
Comment (fmassa) [inline torch/functional.py]: You also need to add it to `__all__` below

File: torch/functional.py [hunk 2]
```diff
@@ -245,3 +246,23 @@ def maybeSqueeze(tensor):
...
+
+def sum(input, axes, keepdims=False, out=None):
+    if isinstance(axes, collections.Iterable)
+        if a.dim() > 3:
+            if keepdims:
+                for ax in axes:
+                    input = input.sum(ax, keepdims=True)
+            else:
+                for ax in sorted(axes, reverse=True):
+                    input = input.sum(ax)
```
Comment (fmassa) [inline torch/functional.py]: Maybe it would be better to add the `out` argument here instead of an extra copy?

File: torch/functional.py [hunk 3]
```diff
@@ -245,3 +246,23 @@ def maybeSqueeze(tensor):
 
     raise ValueError("both arguments to __matmul__ need to be at least 1D, "
                      "but they are {}D and {}D".format(dim_tensor1, dim_tensor2))
+
+
+def sum(input, axes, keepdims=False, out=None):
+    if isinstance(axes, collections.Iterable)
+        if a.dim() > 3:
+            if keepdims:
```
Comment (fmassa) [inline torch/functional.py]: Do we need those two conditions? Can't we just pass the `keepdims` argument, and iterate with `sorted`?

File: torch/functional.py [hunk 4]
```diff
@@ -245,3 +246,23 @@ def maybeSqueeze(tensor):
...
+                    input = input.sum(ax)
+            if out is None:
+                return input
+            else:
+                return out.copy_(input)
+        else:
+            # permute
+            # reduce single dim
+    else:
+        return torch._C.sum(input, axes, keepdims, out)
```
Comment (fmassa) [inline torch/functional.py]: Does this return the `out` tensor here?

**Review by fmassa — COMMENTED** [review]
Thanks for the PR!
I wonder if it would be better to remove the implementation from cwrap, to avoid conflicts? I think that it's better than implementing these ops in cwrap, because we automatically have support for autograd.
Also, what is the behavior of numpy for operations like median when multiple axis are passed? Does it perform multiple kernel calls, or does it transpose+view+kernel call? For `sum` it moght not matter, but for other ops that might make a difference.

File: torch/functional.py (hunk 1 above)
Comment (vlasenkov) [inline torch/functional.py]: of course

File: torch/functional.py (hunk 3 above)
Comment (vlasenkov) [inline torch/functional.py]: yes, we can

File: torch/functional.py (hunk 2 above)
Comment (vlasenkov) [inline torch/functional.py]: Here size of the tensor changes with each iteration, so it anyway requires to allocate new memory every time. In this case `out` won't do the job unless I don't know about some advanced features of Pytorch memory management...

File: torch/functional.py (hunk 4 above)
Comment (vlasenkov) [inline torch/functional.py]: Yes, it does

**Comment by karandwivedi42** [comment]
@zou3519 closed in https://github.com/pytorch/pytorch/pull/6152?

**Comment by tstandley** [comment]
I also want to do var over many axes. is this solved? (same question as @bernardohenz except with var).

I should note that numpy supports this, and the only way to do this in pytorch currently is to compute the mean, subtract (using expand), square, and then take the mean. Basically manually.

**Comment by soumith** [comment]
@tstandley we are working on mean, variance and stdv on multiple axes. @colesbury should put up a PR soon for it.

**Comment by anilkeshwani** [comment]
For signposting: See Reductions tracking issue #61417

**Pull request closed without merging**